### ISSUE TITLE
Parameterized types are not converted when used in union with `str` like `str | list[str]`

### ISSUE THREAD

**Issue by pegna12** [robotframework/robotframework#4809]
After upgrading to Robot 6.1, I encounter unexpected type changes which seem to be linked to [Argument conversion changes](https://github.com/robotframework/robotframework/blob/master/doc/releasenotes/rf-6.1.rst#argument-conversion-changes)

python --version
>>Python 3.10.11

Please consider the following code:

**.robot file:**
`keyword        {"key1":"val1", "key2": ["val2", "val3"]}`

**keywords.py:**
```
import logging
LOGGER = logging.getLogger(__name__)
@keyword
def keyword(events: dict[str, str | list[str]]):
     LOGGER.info(type(events["key2"]))
     return None
```

with Robot 6.0:
>> [info] <class 'list'>

with Robot 6.1:
>> [info] <class 'str'>

Note that if I change the type hint to "list[str] | str" instead of "str | list[str]", the result is the same as 6.0 (i.e list type).
Obviously, this brings other problems when trying to access the dict values afterwards

Could you please advise? I don't think it is intended to convert a list to a str

Thanks

**Comment by pekkaklarck** [comment]
I was able to reproduce the issue using the example in the description and also with this simpler example:

```python
def keyword(arg: str | list[str]):
    print(type(arg))
```
```
${list} =    Create List    foo    bar
Keyword    ${list}
```

This issue is caused by #4705. Commit 72cde3b519ca3547dd24141f33a10e5e87a88cba changed the behavior so that if a type hint is parameterized like `list[str]`, nested items are converted also when the given value is an object, not a string. The problem is that we want to always convert nested items and don't look do they already have correct types. This doesn't cause issue if an argument has a single type hint like `arg: list[str]`, but if it has multiple like `arg: str | list[str]` and none of them can be used without conversion, conversion is attempted one-by-one starting from the first one. In the previous example `str` is first and thus the list ends up being converted to a string.

Fixing this requires changing the logic of checking is conversion needed or not so that nested items are checked separately. That's not too complicated but adds some more code to the already pretty complicated conversion logic. The same issue occurs with all parameterizable types we support (lists, dicts, tuples, sets) and needs to be fixed with all of them. The fix has a benefit that it avoids unnecessary conversions and recreating the container element if items already have correct types. As I [commented](https://github.com/robotframework/robotframework/issues/4705#issuecomment-1489979334) #4705, that would have been a better design to begin with.

This is a regression and needs to thus be fixed in RF 6.1.1.

**Comment by pekkaklarck** [comment]
This issue doesn't occur with other types than `str` because if you have `arg: int | list[int]` and pass it a list of integers, `int` conversion fails and `list[int]` conversion is attempted.

**Comment by pegna12** [comment]
Thank you for the analysis @pekkaklarck  !

**Comment by pekkaklarck** [comment]
A workaround for this issue is having `str` last in the union like `list[str] | str`. Because conversions are attempted from lest to right and `str` conversion always succeeds, having `str` last is a good idea in general. Anyway, the issue is now fixed and starting from RF 6.1.1, containers having correct nested types aren't anymore converted unnecessarily.
